### PR TITLE
Adding component architecture and rework html editor BNN-179 #resolves

### PR DESCRIPTION
**Pull request by nurv**

### CODE REVIEW

File: bennu-toolkit/src/main/webapp/bennu-toolkit/js/localizedString.js [hunk 1]
```diff
@@ -54,7 +54,7 @@
         }
 
         val = val[tag]
-        if (val !== inputField.val()){
+        if (locale !== widget.data("locale")){
```
Comment (pedrosan7os) [inline bennu-toolkit/src/main/webapp/bennu-toolkit/js/localizedString.js]: This does not belong in this pull request

File: bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/Component.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        return COMPONENTS.values();
+    }
+
+    public static String process(String origin) {
+        Document doc = Jsoup.parse(origin);
+        Elements components = doc.select("[bennu-component]");
+
+        for (Element component : components) {
+            String key = component.attr("bennu-component");
+            Element result = COMPONENTS.get(key).process(component);
```
Comment (jcarvalho) [inline bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/Component.java]: You should gracefully handle a definition of a missing component. This way it will just NPE.

File: bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/Component.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        try {
+            COMPONENTS.put(annotation.key(), (Component) type.newInstance());
+        } catch (InstantiationException | IllegalAccessException e) {
+            LOGGER.error("Error while instancing a toolkit component", e);
+        }
+    }
+
+    public static LocalizedString process(LocalizedString origin) {
+        LocalizedString result = new LocalizedString();
```
Comment (jcarvalho) [inline bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/Component.java]: You should use a `LocalizedString.Builder`, as it is a more efficient way to instantiate the multiple locales of a `LocalizedString`.

File: bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/ToolkitResources.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+
+@Path("/bennu-toolkit/")
+public class ToolkitResources extends BennuRestResource {
+    @GET
+    @Path("components")
+    @Produces({ MediaType.APPLICATION_JSON })
+    public Response components() throws IOException {
+        JsonArray array = new JsonArray();
```
Comment (jcarvalho) [inline bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/ToolkitResources.java]: As components are registered on startup, perhaps this response can be cached?

**Comment by jcarvalho** [comment]
Please squash your commits and properly format your POM.

File: bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/Component.java (hunk 3 above)
Comment (nurv) [inline bennu-toolkit/src/main/java/org/fenixedu/bennu/toolkit/components/Component.java]: In this case is more efficient this way, because using the builder I would have to create a localised string for each Locale.

**Comment by nurv** [comment]
POM reformated and most of the comments resolved, just waiting for @pedrosan7os to weigh in.

**Comment by pedrosan7os** [comment]
Integrated in 5ca1752ee7e5a4a863595eae0691f359d36cec18
